cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -611,6 +611,110 @@ class AzureFileSystem::Impl {
+                                 container_client.GetUrl(),
+                                 "): ", response.RawResponse->GetReasonPhrase(), " (",
+                                 static_cast<int>(response.RawResponse->GetStatusCode()),
+                                 "): ", body_text);
+        }
+      } catch (const Azure::Storage::StorageException& exception) {
+        return internal::ExceptionToStatus(
+            "Failed to create a container: " + path.container + " (" +
+                container_client.GetUrl() + ")",
+            exception);
+      }
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto hierarchical_namespace_enabled,
+                          hierarchical_namespace_.Enabled(path.container));
+    if (!hierarchical_namespace_enabled) {
+      return Status::NotImplemented(
+          "Cannot create a directory without hierarchical namespace: ", path.full_path);
+    }
+    auto directory_client = datalake_service_client_->GetFileSystemClient(path.container)
+                                .GetDirectoryClient(path.path_to_file);
+    try {
+      auto response = directory_client.Create();
+      if (response.Value.Created) {
+        return Status::OK();
+      } else {
+        const auto& body = response.RawResponse->GetBody();
+        std::string_view body_text(reinterpret_cast<const char*>(body.data()),
+                                   body.size());
+        return Status::IOError("Failed to create a directory: ", path.path_to_file, " (",
+                               directory_client.GetUrl(),
+                               "): ", response.RawResponse->GetReasonPhrase(), " (",
+                               static_cast<int>(response.RawResponse->GetStatusCode()),
+                               "): ", body_text);
+      }
+    } catch (const Azure::Storage::StorageException& exception) {
+      return internal::ExceptionToStatus(
+          "Failed to create a directory: " + path.path_to_file + " (" +
+              directory_client.GetUrl() + ")",
+          exception);
+    }
+  }
+
+  Status CreateDirRecursive(const AzurePath& path) {
+    if (path.container.empty()) {
+      return Status::Invalid("Cannot create an empty container");
+    }
+
+    auto container_client = blob_service_client_->GetBlobContainerClient(path.container);
+    try {
+      container_client.CreateIfNotExists();
+    } catch (const Azure::Storage::StorageException& exception) {
+      return internal::ExceptionToStatus(
+          "Failed to create a container: " + path.container + " (" +
+              container_client.GetUrl() + ")",
+          exception);
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto hierarchical_namespace_enabled,
+                          hierarchical_namespace_.Enabled(path.container));
+    if (!hierarchical_namespace_enabled) {
+      return Status::NotImplemented(
+          "Cannot create a directory without hierarchical namespace: ", path.full_path);
```

Comment:
I would write "...without hierarchical namespace *support*" so the user is less likely to interpret this as some missing part of the path passed as argument.

Reply:
Thanks! Wording suggestion is always welcome for me because I'm not a native English speaker!
I'll add "support".

Reply:
> Thanks! Wording suggestion is always welcome for me because I'm not a native English speaker!

No worries. This could happen to native English speakers as well. :)